src/message.ts

```diff
@@ -613,6 +625,40 @@ export class Message implements Sayable {
   // }
 
   /**
+   * @param sendTo UserId/RoomId/Room/Contact
+   */
+  public forward(room: Room): Promise<any>
+  public forward(contact: Contact): Promise<any>
+  public forward(id: string): Promise<any>
+  public forward(sendTo: string|Room|Contact): Promise<any> {
+    const m = <MsgRawObj>this.rawObj
+    const newMsg = <MsgRawObj>{}
+    let id = ''
+    if (sendTo instanceof Room || sendTo instanceof Contact) {
+      id = sendTo.id
+    } else {
+      id = sendTo
+    }
+    newMsg.ToUserName = id
+    newMsg.FromUserName = config.puppetInstance().userId || ''
+    newMsg.isTranspond = true
+    newMsg.MsgIdBeforeTranspond = m.MsgIdBeforeTranspond || m.MsgId
+    newMsg.MMSourceMsgId = m.MsgId
+    newMsg.sendByLocal = false
+    newMsg.Content = UtilLib.unescapeHtml(m.Content.replace(/^@\w+:<br\/>/, ''))
+    newMsg.MMActualSender = config.puppetInstance().userId || ''
+    if (m.MMSendContent)
+      newMsg.MMSendContent = m.MMSendContent.replace(/^@\w+:\s/, '')
+    if (m.MMDigest)
+      newMsg.MMDigest = m.MMDigest.replace(/^@\w+:/, '')
+    if (m.MMActualContent)
+      newMsg.MMActualContent = UtilLib.stripHtml(m.MMActualContent.replace(/^@\w+:<br\/>/, ''))
+
+    return config.puppetInstance()
+      .forward(m, newMsg)
```

Comment:
This design is needed

**transpondDialogController** in WxApp:
``` javascript
  angular.module("Controllers").controller("transpondDialogController", ["$rootScope", "$scope", "$timeout", "$state", "$log", "$document", "chatFactory", "contactFactory", "appFactory", "chatroomFactory", "confFactory", "mmpop", "ngDialog", "utilFactory", "stateManageService", "accountFactory",
    function(e, t, a, n, i, o, chatFactory, c, s, l, confFactory, f, u, m, g, accountFactory) {
      function h(msg, userName) {
        if (e.MsgType != confFactory.MSGTYPE_SYS) {
          var newMsg = angular.copy(msg);
          newMsg.ToUserName = userName;
          newMsg.FromUserName = accountFactory.getUserName();
          newMsg.isTranspond = !0;
          newMsg.MsgIdBeforeTranspond = msg.MsgIdBeforeTranspond || msg.MsgId;
          newMsg._h = void 0;
          newMsg._offsetTop = void 0;
          newMsg.MMSourceMsgId = msg.MsgId;
          newMsg.Scene = 2;
          newMsg = chatFactory.createMessage(newMsg); // call createMessage()

          // The following parameters need to be overridden after calling createMessage()
          newMsg.sendByLocal = !1;
          newMsg.Content = m.htmlDecode(newMsg.Content.replace(/^@\w+:<br\/>/, ""));
          newMsg.MMActualSender = accountFactory.getUserName();
          newMsg.MMSendContent && (newMsg.MMSendContent = newMsg.MMSendContent.replace(/^@\w+:\s/, ""));
          newMsg.MMDigest && (newMsg.MMDigest = newMsg.MMDigest.replace(/^@\w+:/, ""));
          newMsg.MMActualContent && (newMsg.MMActualContent = m.clearHtmlStr(newMsg.MMActualContent.replace(/^@\w+:<br\/>/, "")));

          // append and send message
          chatFactory.appendMessage(newMsg);
          chatFactory.sendMessage(newMsg)
        }
      }
```


`**chatFactory.createMessage()**`  in WxApp:
```javascript
createMessage: function(e) { // 创建消息体
  switch (e.FromUserName || (e.FromUserName = accountFactory.getUserName()),
    e.ToUserName || (e.ToUserName = this.getCurrentUserName()), //
    e.ClientMsgId = e.LocalID = e.MsgId = (utilFactory.now() + Math.random().toFixed(3)).replace(".", ""), //生成本地消息id
    e.CreateTime = Math.round(utilFactory.now() / 1e3),
    e.MMStatus = confFactory.MSG_SEND_STATUS_READY, //标记消息的状态为准备状态
    e.sendByLocal = !0, // 是否是本地信息，如果需要转发文件，必须为 false
    e.MsgType) { //判断消息类型

    case confFactory.MSGTYPE_TEXT:
      var t = [];
      e.Content = e.Content.replace(/<input.*?un="(.*?)".*?value="(.*?)".*?>/g,
          function(e, a, n) {
            return t.push(a), n
          }),
        e.MMAtContacts = t.join(","),
        e.MMSendContent = utilFactory.htmlDecode(utilFactory.clearHtmlStr(e.Content.replace(/<(?:img|IMG).*?text="(.*?)".*?>/g,
          function(e, t) {
            return t.replace(confFactory.MM_EMOTICON_WEB, "")
          }).replace(/<(?:br|BR)\/?>/g, "\n"))).replace(/<(.*?)>/g, function(e) {
          return emojiFactory.EmojiCodeMap[emojiFactory.QQFaceMap[e]] || e
        }),
        e.Content = e.Content.replace(/<(?!(img|IMG|br|BR))[^>]*>/g, "").replace(/\n/g, "<br>");
      break;

    case confFactory.MSGTYPE_APP:
      if (e.AppMsgType == confFactory.APPMSGTYPE_URL) break;
      e.AppMsgType = confFactory.APPMSGTYPE_ATTACH,
        e.Content = "<msg><appmsg appid='wxeb7ec651dd0aefa9' sdkver=''><title>" + e.FileName + "</title><des></des><action></action><type>" + confFactory.APPMSGTYPE_ATTACH + "</type><content></content><url></url><lowurl></lowurl><appattach><totallen>" + e.FileSize + "</totallen><attachid>#MediaId#</attachid><fileext>" + (e.MMFileExt || e.MMAppMsgFileExt) + "</fileext></appattach><extinfo></extinfo></appmsg></msg>"
  }
  return e
},
```


***foraward()** in wechaty-bro

```
  function forward(baseData, patchData) {
    var chatFactory = WechatyBro.glue.chatFactory
    var confFactory = WechatyBro.glue.confFactory

    if (!chatFactory || !confFactory) {
      log('forward() chatFactory or confFactory not exist.')
      return false
    }

    try {
      var m = chatFactory.createMessage(baseData)

      // Need to override the parametes after called createMessage()
      m = Object.assign(m, patchData)

      chatFactory.appendMessage(m)
      chatFactory.sendMessage(m)
    } catch (e) {
      log('forward() exception: ' + e.message)
      return false
    }
    return true
  }
```